### PR TITLE
[Merged by Bors] - Unique WorldId

### PR DESCRIPTION
**Pull request by DJMcNab**
# Objective

Fixes these issues:
- `WorldId`s currently aren't necessarily unique
    - I want to guarantee that they're unique to safeguard my librarified version of https://github.com/bevyengine/bevy/discussions/2805
    - There probably hasn't been a collision yet, but they could technically collide
- `SystemId` isn't used for anything
  - It's no longer used now that `Locals` are stored within the `System`.
- `bevy_ecs` depends on rand

## Solution

- Instead of randomly generating `WorldId`s, just use an incrementing atomic counter, panicing on overflow.
- Remove `SystemId` 
    - We do need to allow Locals for exclusive systems at some point, but exclusive systems couldn't access their own `SystemId` anyway.
- Now that these don't depend on rand, move it to a dev-dependency

## Todo

Determine if `WorldId` should be `u32` based instead

### CODE REVIEW

File: crates/bevy_ecs/src/world/mod.rs [hunk 1]
```diff
@@ -22,15 +22,24 @@ use crate::{
...
+
 impl Default for WorldId {
     fn default() -> Self {
-        WorldId(rand::random())
+        let id = MAX_WORLD_ID
+            // We use `Relaxed` here since this atomic only needs to be consistent with itself
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |val| {
+                val.checked_add(1)
+            })
+            .expect("More bevy `World`s have been created than is supported. Please tell us what you were doing, because frankly I'm impressed.");
```
Comment (DJMcNab) [inline crates/bevy_ecs/src/world/mod.rs:41]: I thought I'd removed this second sentence before committing. I'm happy to keep it as-is though 😆

**Review by alice-i-cecile — APPROVED** [review]
Once the error message is cleaned up this LGTM.

I don't have strong feelings about u32 vs u64 for world ids. `u32` is probably slightly more reasonable as there's effectively no way to break those limits without going OOM elsewhere. But on the other hand, there's no real cost to storing more data here 🤔

File: crates/bevy_ecs/src/schedule/run_criteria.rs [hunk 2]
```diff
@@ -407,7 +406,7 @@ impl Default for RunOnce {
     fn default() -> Self {
         Self {
             ran: false,
-            system_id: SystemId::new(),
+
```
Comment (IceSentry) [inline crates/bevy_ecs/src/schedule/run_criteria.rs:409]: Unnecessary whitespace, I'm surprised rustfmt doesn't check for that

**Review by IceSentry — APPROVED** [review]

File: crates/bevy_ecs/src/schedule/run_criteria.rs (hunk 2 above)
Comment (DJMcNab) [reply crates/bevy_ecs/src/schedule/run_criteria.rs]: Thanks!

**Comment by inodentry** [comment]
Regarding `u64` vs `u32` for `WorldId`, I *do* have strong feelings. This is an atomic variable, which means that it relies on appropriate support for atomics from the underlying CPU architecture. There are many platforms that do not support 64-bit atomics, but do support 32-bit atomics. Most 32-bit architectures / compilation targets are like that. If we require 64-bit atomics for bevy, we are restricting portability to such platforms/architectures.

As @alice-i-cecile said, it is unlikely to break the limit of `u32`. It is very doubtful that someone would want to have more than ~4.7 billion `World`s.

Whereas portability and platform support is an actual problem with actual consequences. Let's not needlessly limit the kinds of devices that bevy can run on.

**Comment by DJMcNab** [comment]
Hmm, I hadn't considered that - I wonder if there should be a clippy lint for that (beyond just using the more general disallowed function)

We should use `AtomicUsize` then.

**Comment by cart** [comment]
I dig these changes. I do think that ultimately we might need a unique SystemId for "manually triggered systems" and "reactions", but I'm happy to just re-add that if/when its actually needed. SystemId has been dead code since bevy_ecs's inception. Removing it makes sense.

**Comment by cart** [comment]
Lol I have a nasty habit of  not checking CI before bors-ing. You should just swap out the `WorldId::new()` constructor for a Default impl.

**Comment by DJMcNab** [comment]
That's annoying - I even knew that lint would trigger but forgot to deal with it.

I explicitly chose not to make it a `Default` impl, because it doesn't *really* make sense to have a `WorldId` seperated from a `World`. Additionally, the default value isn't really a default as such imo

I still expose the constructor function publicly because it could be used e.g. for tests in external crates, although I'm not certain that's really needed.

**Review by TheRawMeatball — APPROVED** [review]
